File: 0394-decode-string/0394-decode-string.py
class Solution:
    def decodeString(self, s):
        stack = []
        curr_num = 0
        curr_str = ""

        for ch in s:

            # If digit
            if ch.isdigit():
                curr_num = curr_num * 10 + int(ch)

            # Opening bracket
            elif ch == '[':
                stack.append((curr_str, curr_num))
                curr_str = ""
                curr_num = 0

            # Closing bracket
            elif ch == ']':
                prev_str, num = stack.pop()
                curr_str = prev_str + curr_str * num

            # Character
            else:
                curr_str += ch

        return curr_str

File: 0394-decode-string/0394-decode-string.java
class Solution {
    public String decodeString(String s) {

        Stack<Integer> numStack = new Stack<>();
        Stack<String> strStack = new Stack<>();

        String currStr = "";
        int currNum = 0;

        for (char ch : s.toCharArray()) {

            // If digit
            if (Character.isDigit(ch)) {
                currNum = currNum * 10 + (ch - '0');
            }

            // Opening bracket
            else if (ch == '[') {
                numStack.push(currNum);
                strStack.push(currStr);

                currNum = 0;
                currStr = "";
            }

            // Closing bracket
            else if (ch == ']') {

                int repeat = numStack.pop();
                String prevStr = strStack.pop();

                StringBuilder temp = new StringBuilder(prevStr);

                for (int i = 0; i < repeat; i++) {
                    temp.append(currStr);
                }

                currStr = temp.toString();
            }

            // Normal character
            else {
                currStr += ch;
            }
        }

        return currStr;
    }
}